File: frontend/lib/services/nghiphep/nghi_phep_service.dart
import 'dart:convert';
import '../../model/nghiphep/nghi_phep.dart';
import '../api/api_service.dart';

class NghiPhepService {
  static const String endpoint = 'nghiphep';
  final ApiService _apiService = ApiService();

  // Lấy tất cả đơn nghỉ phép
  Future<List<NghiPhep>> getAllNghiPhep() async {
    final response = await _apiService.get(endpoint);
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Lấy đơn nghỉ phép theo ID
  Future<NghiPhep?> getNghiPhepById(int maNghiPhep) async {
    try {
      final response = await _apiService.get('$endpoint/$maNghiPhep');
      return NghiPhep.fromJson(json.decode(response.body));
    } catch (e) {
      return null;
    }
  }

  // Lấy đơn nghỉ phép theo nhân viên
  Future<List<NghiPhep>> getNghiPhepByNhanVien(int maNV) async {
    final response = await _apiService.get('$endpoint/nhanvien/$maNV');
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Lấy đơn nghỉ phép theo trạng thái
  Future<List<NghiPhep>> getNghiPhepByTrangThai(String trangThai) async {
    final response = await _apiService.get('$endpoint/trangthai/$trangThai');
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Lấy đơn nghỉ phép theo nhân viên và khoảng thời gian
  Future<List<NghiPhep>> getNghiPhepByNhanVienAndKhoangThoiGian(
      int maNV, DateTime tuNgay, DateTime denNgay) async {
    final tuNgayStr = tuNgay.toIso8601String().split('T')[0];
    final denNgayStr = denNgay.toIso8601String().split('T')[0];
    final response = await _apiService.get(
        '$endpoint/nhanvien/$maNV/khoangthoi?tuNgay=$tuNgayStr&denNgay=$denNgayStr');
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Lấy đơn nghỉ phép theo khoảng thời gian
  Future<List<NghiPhep>> getNghiPhepByKhoangThoiGian(
      DateTime tuNgay, DateTime denNgay) async {
    final tuNgayStr = tuNgay.toIso8601String().split('T')[0];
    final denNgayStr = denNgay.toIso8601String().split('T')[0];
    final response = await _apiService.get(
        '$endpoint/khoangthoi?tuNgay=$tuNgayStr&denNgay=$denNgayStr');
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Lấy đơn nghỉ phép theo nhân viên và loại nghỉ
  Future<List<NghiPhep>> getNghiPhepByNhanVienAndLoai(
      int maNV, String loaiNghi) async {
    final response = await _apiService.get('$endpoint/nhanvien/$maNV/loai/$loaiNghi');
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Lấy đơn nghỉ phép chờ duyệt
  Future<List<NghiPhep>> getNghiPhepChoDuyet(int nguoiDuyet) async {
    final response = await _apiService.get('$endpoint/choduyet/$nguoiDuyet');
    final List<dynamic> data = json.decode(response.body);
    return data.map((json) => NghiPhep.fromJson(json)).toList();
  }

  // Tạo đơn nghỉ phép mới
  Future<NghiPhep> createNghiPhep(NghiPhep nghiPhep) async {
    final response = await _apiService.post(endpoint, body: nghiPhep.toJson());
    return NghiPhep.fromJson(json.decode(response.body));
  }

  // Cập nhật đơn nghỉ phép
  Future<NghiPhep> updateNghiPhep(int maNghiPhep, NghiPhep nghiPhep) async {
    final response = await _apiService.put('$endpoint/$maNghiPhep', body: nghiPhep.toJson());
    return NghiPhep.fromJson(json.decode(response.body));
  }

  // Xóa đơn nghỉ phép
  Future<void> deleteNghiPhep(int maNghiPhep) async {
    await _apiService.delete('$endpoint/$maNghiPhep');
  }

  // Duyệt đơn nghỉ phép
  Future<NghiPhep> duyetNghiPhep(
      int maNghiPhep, int nguoiDuyet, bool duyet, String? ghiChu) async {
    final body = {
      'nguoiDuyet': nguoiDuyet,
      'duyet': duyet,
      'ghiChu': ghiChu ?? '',
    };
    final response = await _apiService.put('$endpoint/$maNghiPhep/duyet', body: body);
    return NghiPhep.fromJson(json.decode(response.body));
  }

  // Lấy số ngày phép đã sử dụng trong năm
  Future<int> getSoNgayPhepDaSuDung(int maNV, {int? nam}) async {
    final year = nam ?? DateTime.now().year;
    final startOfYear = DateTime(year, 1, 1);
    final endOfYear = DateTime(year, 12, 31);
    
    final danhSach = await getNghiPhepByNhanVienAndKhoangThoiGian(maNV, startOfYear, endOfYear);
    final daDuyet = danhSach.where((item) => item.isDuocDuyet).toList();
    
    return daDuyet.fold<int>(0, (sum, item) => sum + (item.soNgay ?? 0));
  }

  // Lấy đơn nghỉ phép chờ duyệt (tất cả)
  Future<List<NghiPhep>> getAllNghiPhepChoDuyet() async {
    return getNghiPhepByTrangThai('CHO_DUYET');
  }
}

File: backend/src/main/java/com/example/worktrack/service/nghiphep/NghiPhepService.java
package com.example.worktrack.service.nghiphep;

import com.example.worktrack.model.nghiphep.NghiPhep;
import com.example.worktrack.repository.nghiphep.NghiPhepRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class NghiPhepService {

    private final NghiPhepRepository nghiPhepRepository;

    public List<NghiPhep> getAllNghiPhep() {
        return nghiPhepRepository.findAll();
    }

    public Optional<NghiPhep> getNghiPhepById(Integer maNghiPhep) {
        return nghiPhepRepository.findById(maNghiPhep);
    }

    public List<NghiPhep> getNghiPhepByNhanVien(Integer maNV) {
        return nghiPhepRepository.findByMaNVAndDaXoa(maNV, false);
    }

    public List<NghiPhep> getNghiPhepByTrangThai(String trangThai) {
        return nghiPhepRepository.findByTrangThaiAndDaXoa(trangThai, false);
    }

    public List<NghiPhep> getNghiPhepByNhanVienAndKhoangThoiGian(
            Integer maNV, LocalDate tuNgay, LocalDate denNgay) {
        return nghiPhepRepository.findByMaNVAndTuNgayBetween(maNV, tuNgay, denNgay);
    }

    public List<NghiPhep> getNghiPhepByKhoangThoiGian(LocalDate tuNgay, LocalDate denNgay) {
        return nghiPhepRepository.findByTuNgayBetween(tuNgay, denNgay);
    }

    public List<NghiPhep> getNghiPhepByNhanVienAndLoai(Integer maNV, String loaiNghi) {
        return nghiPhepRepository.findByMaNVAndLoaiNghi(maNV, loaiNghi);
    }

    public List<NghiPhep> getNghiPhepChoDuyet(Integer nguoiDuyet) {
        return nghiPhepRepository.findByNguoiDuyetAndTrangThai(nguoiDuyet, "CHO_DUYET");
    }

    public NghiPhep createNghiPhep(NghiPhep nghiPhep) {
        nghiPhep.setDaXoa(false);
        nghiPhep.setTrangThai("CHO_DUYET");

        // Tính số ngày nghỉ
        long soNgay = ChronoUnit.DAYS.between(nghiPhep.getTuNgay(), nghiPhep.getDenNgay()) + 1;
        nghiPhep.setSoNgay((int) soNgay);

        return nghiPhepRepository.save(nghiPhep);
    }

    public NghiPhep updateNghiPhep(Integer maNghiPhep, NghiPhep nghiPhep) {
        Optional<NghiPhep> existingNghiPhep = nghiPhepRepository.findById(maNghiPhep);
        if (existingNghiPhep.isPresent()) {
            NghiPhep nghiPhepEntity = existingNghiPhep.get();
            nghiPhepEntity.setMaNV(nghiPhep.getMaNV());
            nghiPhepEntity.setTuNgay(nghiPhep.getTuNgay());
            nghiPhepEntity.setDenNgay(nghiPhep.getDenNgay());
            nghiPhepEntity.setLoaiNghi(nghiPhep.getLoaiNghi());
            nghiPhepEntity.setLyDo(nghiPhep.getLyDo());

            // Tính lại số ngày
            long soNgay = ChronoUnit.DAYS.between(nghiPhep.getTuNgay(), nghiPhep.getDenNgay()) + 1;
            nghiPhepEntity.setSoNgay((int) soNgay);

            return nghiPhepRepository.save(nghiPhepEntity);
        }
        throw new RuntimeException("Không tìm thấy đơn nghỉ phép với mã: " + maNghiPhep);
    }

    public void deleteNghiPhep(Integer maNghiPhep) {
        Optional<NghiPhep> nghiPhep = nghiPhepRepository.findById(maNghiPhep);
        if (nghiPhep.isPresent()) {
            NghiPhep nghiPhepEntity = nghiPhep.get();
            nghiPhepEntity.setDaXoa(true);
            nghiPhepRepository.save(nghiPhepEntity);
        } else {
            throw new RuntimeException("Không tìm thấy đơn nghỉ phép với mã: " + maNghiPhep);
        }
    }

    public NghiPhep duyetNghiPhep(Integer maNghiPhep, Integer nguoiDuyet, boolean duyet, String ghiChu) {
        Optional<NghiPhep> existingNghiPhep = nghiPhepRepository.findById(maNghiPhep);
        if (existingNghiPhep.isPresent()) {
            NghiPhep nghiPhep = existingNghiPhep.get();
            nghiPhep.setNguoiDuyet(nguoiDuyet);
            nghiPhep.setNgayDuyet(LocalDateTime.now());
            nghiPhep.setTrangThai(duyet ? "DA_DUYET" : "TU_CHOI");
            nghiPhep.setGhiChuDuyet(ghiChu);
            return nghiPhepRepository.save(nghiPhep);
        }
        throw new RuntimeException("Không tìm thấy đơn nghỉ phép với mã: " + maNghiPhep);
    }
}